src/matmul_kernels.py

```diff
@@ -0,0 +1,157 @@
+import torch
+import triton
+
+assert triton.__version__.startswith("2.1.0"), f"found triton {triton.__version__}, want 2.1.0*"
+import torch
+import triton.language as tl
+
+
+@triton.autotune(
+    configs=[
+        triton.Config({"UNUSED": 1}, num_stages=num_stages, num_warps=num_warps)
+        for num_stages in (1, 2, 3, 4, 5)
+        for num_warps in (1, 2, 4, 8)
+    ],
+    key=[
+        "in_features",
+        "out_features",
+        "num_codebooks",
+        "codebook_size",
+        "out_group_size",
+        "in_group_size",
+        "num_input_groups",
+        "num_input_groups_next_power_of_2",
+        "compute_in_fp32",
+    ],
+)
+@triton.jit
+def _aqlm_gemv_simple(
+    input_vec_ptr,
+    output_vec_ptr,
+    codes_i16_ptr,
+    codebooks_ptr,
+    scales_ptr,
+    in_features: tl.constexpr,
+    out_features: tl.constexpr,
+    num_codebooks: tl.constexpr,
+    codebook_size: tl.constexpr,
+    out_group_size: tl.constexpr,
+    in_group_size: tl.constexpr,
+    num_input_groups: tl.constexpr,
+    num_input_groups_next_power_of_2: tl.constexpr,
+    compute_in_fp32: tl.constexpr,
+    UNUSED: tl.constexpr,
+):
+    # variables ending with "_i" mean "for i-th output unit"
+    pid = tl.program_id(axis=0)  # [0, 1, ... {out_features-1}]
+
+    # Stage 1: load input data
+    input_vec = tl.load(
+        input_vec_ptr
+        + tl.arange(0, num_input_groups_next_power_of_2)[:, None, None] * in_group_size
+        + tl.arange(0, in_group_size)[None, None, :],
+        mask=tl.arange(0, num_input_groups_next_power_of_2)[:, None, None] < num_input_groups,
+    )
+    # [in_features//in_group_size, 1, group_size]
+    # Note: we could simply load input_vec then reshape
+    #     input_vec = tl.load(input_vec_ptr + tl.arange(0, in_features))  # [in_features]
+    #     input_vec = tl.view(input_vec, [num_input_groups, 1, in_group_size])
+    #     , but this does not work because tl.view may reorder elements arbitrarily; see its docstring
+
+    # Stage 2: load integer codes for the active row
+    # [in_features // in_group_size, num_codebooks]
+    codes_i_ptrs = (
+        codes_i16_ptr
+        + pid * num_input_groups * num_codebooks
+        + tl.arange(0, num_input_groups_next_power_of_2)[:, None] * num_codebooks
+        + tl.arange(0, num_codebooks)[None, :]
+    )
+    codes_i_mask_1d = tl.arange(0, num_input_groups_next_power_of_2) < num_input_groups
+
+    codes_i = tl.load(codes_i_ptrs, mask=codes_i_mask_1d[:, None])  # [in_features//in_group_size, num_codebooks]
+    if codes_i.dtype == tl.int16:
```

Comment:
The packing strategy handles negative codes for int32, but the kernel only checks for int16

Reply:
I'll just remove the kernel from this PR. It would be a matter for another PR.

Reply:
I brought it back